server/CGameHandler.cpp

```diff
@@ -2962,6 +2965,262 @@ void CGameHandler::load(const std::string & filename)
+			freeSlots.pop();
+			destMap.insert(lb, TMapCreatureSlot::value_type(currCreature, currFreeSlot));
+		}
+		moves.insert(std::make_pair(currSlot, std::make_pair(destMap[currCreature], quantity)));
+		slotsLeft--;
+	}
+	if(slotsLeft == 1)
+	{
+		auto lastCreature = setSrc.getCreature(srcSlot);
+		auto slotToMove = SlotID();
+		// Try to find a slot for last creature
+		if(destMap.find(lastCreature) == destMap.end())
+		{
+			if(!freeSlots.empty())
+				slotToMove = freeSlots.front();
+		}
+		else
+		{
+			slotToMove = destMap[lastCreature];
+		}
+
+		if(slotToMove != SlotID())
+		{
+			const bool needsLastStack = armySrc->needsLastStack();
+			const auto quantity = setSrc.getStackCount(srcSlot) - (needsLastStack ? 1 : 0);
+			moves.insert(std::make_pair(srcSlot, std::make_pair(slotToMove, quantity)));
+		}
+	}
+	BulkRebalanceStacks bulkRS;
+
+	for(auto & move : moves)
+	{
+		RebalanceStacks rs;
+		rs.srcArmy = armySrc->id;
+		rs.dstArmy = armyDest->id;
+		rs.srcSlot = move.first;
+		rs.dstSlot = move.second.first;
+		rs.count = move.second.second;
+		bulkRS.moves.push_back(rs);
+	}
+	sendAndApply(&bulkRS);
+	return true;
+}
+
+bool CGameHandler::bulkSmartSplitStack(SlotID slotSrc, ObjectInstanceID srcOwner)
+{
+	if(!slotSrc.validSlot() && complain(complainInvalidSlot))
+		return false;
+
+	const CArmedInstance * army = static_cast<const CArmedInstance*>(getObjInstance(srcOwner));
+	const CCreatureSet & creatureSet = *army;
+
+	if(!vstd::contains(creatureSet.stacks, slotSrc) && complain(complainNoCreatures))
+		return false;
+
+	auto actualAmount = creatureSet.getStackCount(slotSrc);
+
+	if(actualAmount <= 1 && complain(complainNoCreatures))
+		return false;
+
+	auto freeSlot = creatureSet.getFreeSlot();
+	auto currentCreature = creatureSet.getCreature(slotSrc);
+
+	if(freeSlot == SlotID() && creatureSet.isCreatureBalanced(currentCreature))
+		return true;
+
+	auto creatureSlots = creatureSet.getCreatureSlots(currentCreature, SlotID(-1), 1); // Ignore slots where's only 1 creature, don't ignore slotSrc
+	auto totalCreatureSlots = creatureSlots.size();
+	TQuantity totalCreatures = 0;
+
+	for(auto slot : creatureSlots)
+		totalCreatures += creatureSet.getStackCount(slot);
+
+	if(freeSlot != SlotID())
+	{
+		creatureSlots.push_back(freeSlot);
+		totalCreatureSlots++;
+	}
+
+	if(!totalCreatureSlots && complain("No available slots for smart rebalancing"))
+		return false;
+
+	if(totalCreatures <= 1 && complain("Total creatures number is invalid"))
+		return false;
+
+	auto rem = totalCreatures % totalCreatureSlots;
+	auto quotient = totalCreatures / totalCreatureSlots;
+
+	// totalCreatures == rem * (quotient + 1) + (totalCreatureSlots - rem) * quotient;
+	// Proof: r(q+1)+(s-r)q = rq+r+qs-rq = r+qs = total, where total/s = q+r/s
+
+	BulkSmartRebalanceStacks bulkSRS;
+
+	if(freeSlot != SlotID())
+	{
+		RebalanceStacks rs;
+		rs.srcArmy = rs.dstArmy = army->id;
+		rs.srcSlot = slotSrc;
+		rs.dstSlot = freeSlot;
+		rs.count = 1;
```

Comment:
Why 1? What does it mean?

Reply:
Just fill free slot (if exists) with target creature, then adjust count.